File: Python/OneLang/One/Ast/References.py
from OneLangStdLib import *
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Statements as stats
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Interfaces as ints

class Reference(exprs.Expression):
    def __init__(self):
        super().__init__()

class VariableReference(Reference):
    def __init__(self):
        super().__init__()
    
    def get_variable(self):
        raise Error("Abstract method")

class ClassReference(Reference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.class_references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        raise Error("ClassReference cannot have a type!")

class GlobalFunctionReference(Reference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        raise Error("GlobalFunctionReference cannot have a type!")
    
    def get_method_base(self):
        return self.decl

class MethodParameterReference(VariableReference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        super().set_actual_type(type, False, ArrayHelper.some(lambda x: astTypes.TypeHelper.is_generic(x.type), self.decl.parent_method.parameters) if isinstance(self.decl.parent_method, types.Lambda) else len(self.decl.parent_method.parent_class.type_arguments) > 0 if isinstance(self.decl.parent_method, types.Constructor) else len(self.decl.parent_method.type_arguments) > 0 or len(self.decl.parent_method.parent_interface.type_arguments) > 0 if isinstance(self.decl.parent_method, types.Method) else False)
    
    def get_variable(self):
        return self.decl

class EnumReference(Reference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        raise Error("EnumReference cannot have a type!")

class EnumMemberReference(Reference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        if not (isinstance(type, astTypes.EnumType)):
            raise Error("Expected EnumType!")
        super().set_actual_type(type)

class StaticThisReference(Reference):
    def __init__(self, cls_):
        self.cls_ = cls_
        super().__init__()
        cls_.static_this_references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        raise Error("StaticThisReference cannot have a type!")

class ThisReference(Reference):
    def __init__(self, cls_):
        self.cls_ = cls_
        super().__init__()
        cls_.this_references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        if not (isinstance(type, astTypes.ClassType)):
            raise Error("Expected ClassType!")
        super().set_actual_type(type, False, len(self.cls_.type_arguments) > 0)

class SuperReference(Reference):
    def __init__(self, cls_):
        self.cls_ = cls_
        super().__init__()
        cls_.super_references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        if not (isinstance(type, astTypes.ClassType)):
            raise Error("Expected ClassType!")
        super().set_actual_type(type, False, len(self.cls_.type_arguments) > 0)

class VariableDeclarationReference(VariableReference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def get_variable(self):
        return self.decl

class ForVariableReference(VariableReference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def get_variable(self):
        return self.decl

class CatchVariableReference(VariableReference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def get_variable(self):
        return self.decl

class ForeachVariableReference(VariableReference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.references.append(self)
    
    def get_variable(self):
        return self.decl

class StaticFieldReference(VariableReference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.static_references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        if astTypes.TypeHelper.is_generic(type):
            raise Error("StaticField's type cannot be Generic")
        super().set_actual_type(type)
    
    def get_variable(self):
        return self.decl

class StaticPropertyReference(VariableReference):
    def __init__(self, decl):
        self.decl = decl
        super().__init__()
        decl.static_references.append(self)
    
    def set_actual_type(self, type, allow_void = False, allow_generic = False):
        if astTypes.TypeHelper.is_generic(type):
            raise Error("StaticProperty's type cannot be Generic")
        super().set_actual_type(type)
    
    def get_variable(self):
        return self.decl

class InstanceFieldReference(VariableReference):
    def __init__(self, object, field):
        self.object = object
        self.field = field
        super().__init__()
        field.instance_references.append(self)
    
    def get_variable(self):
        return self.field

class InstancePropertyReference(VariableReference):
    def __init__(self, object, property):
        self.object = object
        self.property = property
        super().__init__()
        property.instance_references.append(self)
    
    def get_variable(self):
        return self.property

File: PHP/src/OneLang/One/Ast/References.php
<?php

namespace OneLang\One\Ast\References;

use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Enum;
use OneLang\One\Ast\Types\MethodParameter;
use OneLang\One\Ast\Types\GlobalFunction;
use OneLang\One\Ast\Types\Field;
use OneLang\One\Ast\Types\Property;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Types\EnumMember;
use OneLang\One\Ast\Types\IMethodBase;
use OneLang\One\Ast\Types\Lambda;
use OneLang\One\Ast\Types\Constructor;
use OneLang\One\Ast\Types\IVariable;
use OneLang\One\Ast\Statements\VariableDeclaration;
use OneLang\One\Ast\Statements\ForVariable;
use OneLang\One\Ast\Statements\ForeachVariable;
use OneLang\One\Ast\Statements\CatchVariable;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\TypeRestriction;
use OneLang\One\Ast\AstTypes\EnumType;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\TypeHelper;
use OneLang\One\Ast\Interfaces\IExpression;
use OneLang\One\Ast\Interfaces\IType;

interface IReferencable {
    function createReference();
}

interface IGetMethodBase {
    function getMethodBase();
}

class Reference extends Expression {
    
}

class VariableReference extends Reference {
    function getVariable() {
        throw new \OneCore\Error("Abstract method");
    }
}

class ClassReference extends Reference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->classReferences[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        throw new \OneCore\Error("ClassReference cannot have a type!");
    }
}

class GlobalFunctionReference extends Reference implements IGetMethodBase {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        throw new \OneCore\Error("GlobalFunctionReference cannot have a type!");
    }
    
    function getMethodBase() {
        return $this->decl;
    }
}

class MethodParameterReference extends VariableReference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        parent::setActualType($type, false, $this->decl->parentMethod instanceof Lambda ? \OneCore\ArrayHelper::some($this->decl->parentMethod->parameters, function ($x) { return TypeHelper::isGeneric($x->type); }) : ($this->decl->parentMethod instanceof Constructor ? count($this->decl->parentMethod->parentClass->typeArguments) > 0 : ($this->decl->parentMethod instanceof Method ? count($this->decl->parentMethod->typeArguments) > 0 || count($this->decl->parentMethod->parentInterface->typeArguments) > 0 : false)));
    }
    
    function getVariable() {
        return $this->decl;
    }
}

class EnumReference extends Reference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        throw new \OneCore\Error("EnumReference cannot have a type!");
    }
}

class EnumMemberReference extends Reference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        if (!($type instanceof EnumType))
            throw new \OneCore\Error("Expected EnumType!");
        parent::setActualType($type);
    }
}

class StaticThisReference extends Reference {
    public $cls;
    
    function __construct($cls) {
        parent::__construct();
        $this->cls = $cls;
        $cls->staticThisReferences[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        throw new \OneCore\Error("StaticThisReference cannot have a type!");
    }
}

class ThisReference extends Reference {
    public $cls;
    
    function __construct($cls) {
        parent::__construct();
        $this->cls = $cls;
        $cls->thisReferences[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        if (!($type instanceof ClassType))
            throw new \OneCore\Error("Expected ClassType!");
        parent::setActualType($type, false, count($this->cls->typeArguments) > 0);
    }
}

class SuperReference extends Reference {
    public $cls;
    
    function __construct($cls) {
        parent::__construct();
        $this->cls = $cls;
        $cls->superReferences[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        if (!($type instanceof ClassType))
            throw new \OneCore\Error("Expected ClassType!");
        parent::setActualType($type, false, count($this->cls->typeArguments) > 0);
    }
}

class VariableDeclarationReference extends VariableReference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function getVariable() {
        return $this->decl;
    }
}

class ForVariableReference extends VariableReference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function getVariable() {
        return $this->decl;
    }
}

class CatchVariableReference extends VariableReference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function getVariable() {
        return $this->decl;
    }
}

class ForeachVariableReference extends VariableReference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->references[] = $this;
    }
    
    function getVariable() {
        return $this->decl;
    }
}

class StaticFieldReference extends VariableReference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->staticReferences[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        if (TypeHelper::isGeneric($type))
            throw new \OneCore\Error("StaticField's type cannot be Generic");
        parent::setActualType($type);
    }
    
    function getVariable() {
        return $this->decl;
    }
}

class StaticPropertyReference extends VariableReference {
    public $decl;
    
    function __construct($decl) {
        parent::__construct();
        $this->decl = $decl;
        $decl->staticReferences[] = $this;
    }
    
    function setActualType($type, $allowVoid = false, $allowGeneric = false) {
        if (TypeHelper::isGeneric($type))
            throw new \OneCore\Error("StaticProperty's type cannot be Generic");
        parent::setActualType($type);
    }
    
    function getVariable() {
        return $this->decl;
    }
}

class InstanceFieldReference extends VariableReference {
    public $object;
    public $field;
    
    function __construct($object, $field) {
        parent::__construct();
        $this->object = $object;
        $this->field = $field;
        $field->instanceReferences[] = $this;
    }
    
    function getVariable() {
        return $this->field;
    }
}

class InstancePropertyReference extends VariableReference {
    public $object;
    public $property;
    
    function __construct($object, $property) {
        parent::__construct();
        $this->object = $object;
        $this->property = $property;
        $property->instanceReferences[] = $this;
    }
    
    function getVariable() {
        return $this->property;
    }
}
